File: Difficulty/Hard/[146] LRU Cache/solution.cpp
/*
 * @lc app=leetcode id=146 lang=cpp
 *
 * [146] LRU Cache
 */

// @lc code=start
static int __ = [](){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class LRUCache {
public:
    LRUCache(int capacity) : _capacity(capacity) {}
    
    int get(int key) {
        auto it = cache.find(key);
        if (it == cache.end()) return -1;
        touch(it);
        return it->second.first;
    }
    
    void put(int key, int value) {
        auto it = cache.find(key);
        if (it != cache.end()) touch(it);
        else {
            if (cache.size() == _capacity) {
                cache.erase(used.back());
                used.pop_back();
            }
            used.push_front(key);
        }
        cache[key] = {value, used.begin()};
    }

private:
    typedef list<int> LI;
    typedef pair<int, LI::iterator> LLI;
    typedef unordered_map<int, LLI> HILLI;

    void touch(HILLI::iterator it) {
        int key = it->first;
        used.erase(it->second.second);
        used.push_front(key);
        it->second.second = used.begin();
    }

    int _capacity;
    LI used;
    HILLI cache;
};


class LRUCache {
public:
    LRUCache(int capacity) {
        cap = capacity;
        size = 0;
    }
    
    int get(int key) {
        for (auto& p : time_map) 
            time_map[p.first] = p.second + 1;
        if (cache_map.find(key) != cache_map.end()){
            time_map[key] = 0;
            return cache_map[key];
        }else return -1;
    }
    
    void put(int key, int value) {
        if (cache_map.find(key) != cache_map.end()) {
            for (auto& p : time_map) 
                time_map[p.first] = p.second + 1;
            cache_map[key] = value;
            time_map[key] = 0;
            return;
        }
        if (size == cap) {
            int maxtime = 0, maxtime_key = 0;
            for (auto& p : time_map) {
                if (p.second > maxtime){
                    maxtime_key = p.first;
                    maxtime = p.second;
                }
                time_map[p.first] = p.second + 1;
            }
            cache_map.erase(maxtime_key);
            time_map.erase(maxtime_key);
            cache_map[key] = value;
            time_map[key] = 0;
        }else{
            for (auto& p : time_map) 
                time_map[p.first] = p.second + 1;
            cache_map[key] = value;
            time_map[key] = 0;
            size++;
        }
    }

private:
    int size;
    int cap;
    unordered_map<int, int> cache_map;
    unordered_map<int, int> time_map;
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */
// @lc code=end

File: Difficulty/Hard/[146] LRU Cache/solution.py
class Node:
    def __init__(key, value):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = dict()
        self.head = Node(0, 0)
        self.tail = Node(0, 0)
        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, key: int) -> int:
        if key in self.cache:
            n = self.cache[key]
            self._remove(n)
            self._add(n)
            return n.val
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self._remove(self.cache[key])
        n = Node(key, value)
        self._add(n)
        self.cache[key] = n
        if len(self.cache) > self.capacity:
            n = self.head.next
            self._remove(n)
            del self.cache[n.key]
    
    def _remove(self, node):
        p = node.prev
        n = node.next
        p.next = n
        n.prev = p
    
    def _add(self, node):
        p = self.tail.prev
        p.next = node
        node.next = self.tail
        self.tail.prev = node
        node.prev = p

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)